src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -222,6 +228,78 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
         return HandleRequestBodyAndCompileRequestDelegate(responseWritingMethodCall, factoryContext);
     }
 
+    private static void AddTypeProvidedMetadata(MethodInfo methodInfo, FactoryContext factoryContext)
+    {
+        const string argumentNullExceptionMessage = "The IEnumerable<object> returned from GetMetadata must not be null.";
+
+        // Get metadata from parameter types
+        var parameters = methodInfo.GetParameters();
+        foreach (var parameter in parameters)
+        {
+            if (typeof(IProvideEndpointParameterMetadata).IsAssignableFrom(parameter.ParameterType))
+            {
+                // Parameter type implements IProvideEndpointParameterMetadata
+                var metadata = GetMetadataForParameterMethod.MakeGenericMethod(parameter.ParameterType).Invoke(null, new object?[] { parameter, factoryContext.ServiceProvider });
+                if (metadata is null)
+                {
+                    throw new ArgumentNullException(null, argumentNullExceptionMessage);
+                }
+
+                factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+            }
+
+            if (typeof(IProvideEndpointMetadata).IsAssignableFrom(parameter.ParameterType))
+            {
+                // Parameter type implements IProvideEndpointMetadata
+                var metadata = GetMetadataForEndpointMethod.MakeGenericMethod(parameter.ParameterType).Invoke(null, new object?[] { methodInfo, factoryContext.ServiceProvider });
+                if (metadata is null)
+                {
+                    throw new ArgumentNullException(null, argumentNullExceptionMessage);
+                }
+
+                factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+            }
+        }
+
+        // Get metadata from return type
+        if (methodInfo.ReturnType is not null && typeof(IProvideEndpointMetadata).IsAssignableFrom(methodInfo.ReturnType))
+        {
+            // Return type implements IProvideEndpointMetadata
+            var metadata = GetMetadataForEndpointMethod.MakeGenericMethod(methodInfo.ReturnType).Invoke(null, new object?[] { methodInfo, factoryContext.ServiceProvider });
+            if (metadata is null)
+            {
+                throw new ArgumentNullException(null, argumentNullExceptionMessage);
+            }
+
+            factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+        }
+
+        // Get metadata from method attributes that implement IProvideEndpointMetadata
+        var methodAttrs = methodInfo.GetCustomAttributes(inherit: true).OfType<IProvideEndpointMetadata>();
```

Comment:
`inherit: true` is interesting. I wouldn't necessarily expect this, but I don't expect it to come up much. Is there a test verifying we pick up inherited attributes?

Reply:
Based on @davidfowl's comments I'm going to remove the support for this interface on method attributes anyway, so this won't matter for now :smile: